### DISCUSSION TITLE
[Question]: Is it possible to use plugins with 'custom endpoints'??

### DISCUSSION THREAD

**Discussion by bpawnzZ** [danny-avila/LibreChat#2482] [Q&A]
### What is your question?

Specifically I am trying to get search functionality/plugins in general would be nice out of custom endpoints.  Specifically I am using openrouter.  

Thank you

### More Details

Is this functionality not possible?. If it is can someone please point me in the right direction.  

Thank you

### What is the main subject of your question?

LibreChat Plugins

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by bpawnzZ** [comment]
If anyone has any ideas I am all ears?? Like for instance is there a way I can use meilisearch with my custom endpoint?  Just seems like this might be a maybe since it is running in its own container. Idk..  Awesome project though!  This is by far the best chat-ui type application I have thus found.

**Reply by gabcodec** [reply]
I'm wondering the same thing, I'm trying to make it work with my LM Studio api server from other PC in my network. The normal model chat works fine but with the plugins one, I see it includes the plugins as tools in the lm studio logs and even tells the model to use them (I have wolfram alpha and google enabled), with the exception of google and I don't understand why. Furthermore, if I tell it to use any of them it hallucinates like it's using them when in reality it's not.

**Reply by gabcodec** [reply]
I've also noticed that it fails to generate titles, maybe because function calling doesn't work?

**Reply by danny-avila** [reply]
This is not supported yet. Once I get around to it, the custom endpoint would need to be functions capable following OpenAI/Anthropic/Google specs. If there is a standard set by LMStudio, I can investigate that as well.

**Reply by gabcodec** [reply]
I see, that's unfortunate. But, I just wanted to say your project is awesome! I would really appreciate if you could implement it. Thank you and have a great day.

**Reply by bpawnzZ** [reply]
It would likely be not that hard for some models.  Why i say that is I use CrewAI which supports OpenAI compatible endpoints and supports Serpapi for search. And I am using mistralai/mixtral-8x22b-instruct against openrouter's api with crew. Serpapi plugin might be the easiest path to get search against at least some of openrouter's models.  Llama 3 definitly has issues in CrewAI and doesn't work seamlessly like mixtral models, so i do see your concern you stated above.  But yeah.. Just an idea that came to mind.

**Reply by bpawnzZ** [reply]
FYI guys. The workaround for getting plugins with 3rd party models working is getting LiteLLM setup correctly and setting the OpenAI Proxy/environmental variable for LiteLLM's URL. I have mistralai/mixtral-8x22b-instruct via openrouter working perfectly. Like no issues I have spotted this far testing. Getting it to this point took some patience and trial and error but it's working flawlessly.

Just understand this is not officially supported and Devs will not give support on this method. Which is understandable from their perspective and I know they are officially rolling out 3rd party support in the next few months. But for the impatient like myself this does work 


For those interested
docker run command 
````
docker run -d \
    -v $(pwd)/litellm_config.yaml:/app/config.yaml \
    -e LLM_API_KEY=sk-or-v1- \
    -e LLM_API_BASE=https://openrouter.ai/api/v1 \
    -p 4000:4000 \
    -h litellmSRV \
    ghcr.io/berriai/litellm:main-latest \
    --config /app/config.yaml --detailed_debug --model openrouter/mistralai/mixtral-8x22b-instruct
````

litellm_config.yaml
````
model_list:
  - model_name: mixtral-8x22b-instruct
    litellm_params:
      model: openrouter/mistralai/mixtral-8x22b-instruct
      api_base: https://openrouter.ai/api/v1
      api_key: sk-or-
      rpm: 30
````

**Reply by danny-avila** [reply]
@bpawnzZ maybe you can do a blog post and we’ll host it on the LibreChat site?

**Reply by danny-avila** [reply]
https://www.librechat.ai/blog/2024-04-17_blog_guide

**Reply by bpawnzZ** [reply]
> @bpawnzZ maybe you can do a blog post and we’ll host it on the LibreChat site?

I don't mind that man.  No problem.
There are a couple networking things that may not be obvious to people not very into docker. But I'll explain it all in a blog post and try to have it up in next 48 hrs

**Reply by fuegovic** [reply]
@bpawnzZ here's a couple of links to help get started with the blog:
https://www.librechat.ai/blog/2024-04-17_blog_guide
https://github.com/LibreChat-AI/librechat.ai

**Reply by Phlasse** [reply]
> FYI guys. The workaround for getting plugins with 3rd party models working is getting LiteLLM setup correctly and setting the OpenAI Proxy/environmental variable for LiteLLM's URL. I have mistralai/mixtral-8x22b-instruct via openrouter working perfectly. Like no issues I have spotted this far testing. Getting it to this point took some patience and trial and error but it's working flawlessly.
> 
> Just understand this is not officially supported and Devs will not give support on this method. Which is understandable from their perspective and I know they are officially rolling out 3rd party support in the next few months. But for the impatient like myself this does work
> 
> For those interested docker run command
> 
> ```
> docker run -d \
>     -v $(pwd)/litellm_config.yaml:/app/config.yaml \
>     -e LLM_API_KEY=sk-or-v1- \
>     -e LLM_API_BASE=https://openrouter.ai/api/v1 \
>     -p 4000:4000 \
>     -h litellmSRV \
>     ghcr.io/berriai/litellm:main-latest \
>     --config /app/config.yaml --detailed_debug --model openrouter/mistralai/mixtral-8x22b-instruct
> ```
> 
> litellm_config.yaml
> 
> ```
> model_list:
>   - model_name: mixtral-8x22b-instruct
>     litellm_params:
>       model: openrouter/mistralai/mixtral-8x22b-instruct
>       api_base: https://openrouter.ai/api/v1
>       api_key: sk-or-
>       rpm: 30
> ```

@bpawnzZ I am working on the same thing. I get the approach you take, but I am having issues with the proper configuration of the librechat. How did you configure it in the .env file or the librechat.yaml?

**Comment by danny-avila** [comment]
“Agents” are the answer to this

https://github.com/danny-avila/LibreChat/issues/3607
